Spanish Implementation

Laifsyn opened the pull request:
Submitting as draft because
1-) I'm unclear how pull requests actually works
2-) I haven't properly made an Integration Test because I found it extremely un-ergonomic to constantly build the Num2Words settings after every .to_word() call

I also submitted a .devcontainer configuration because I found it extremely helpful to be able to test stuffs in a virtual environment specially so when I didn't have available a beefy-enough computer (I might've forgotten to add Rust Docs Viewer in the extensions though)

Laifsyn committed:
No Deps Working Example with i32

Laifsyn committed:
Staging changes on branch for dev-ing

Laifsyn committed:
Implement Tests for to_cardinal method

Laifsyn committed:
Implement Tests for to_cardinal method

Laifsyn committed:
ammending - eliminate prepended zeros from tests

Laifsyn committed:
Fix weird merging

Laifsyn committed:
More unified test for cardinal conversion

Laifsyn committed:
Add Veinti Flavor and setters for Spanish fields

Laifsyn committed:
Refactoro with BigFloat

Laifsyn committed:
Consume Vec of Triplets instead of reading it

Laifsyn committed:
Refactor out the Integer equivalent of to_cardinal()

Laifsyn committed:
Conversion for fractional number

Laifsyn committed:
Tweak comments and error checking

Laifsyn committed:
Staging: to_ordinal implementation

Laifsyn committed:
defaults to using "veinti..." flavor instead of "veinte y..."

Laifsyn committed:
Stage Changes: to_ordinal progress

Laifsyn committed:
Complete ordinal implementation

Laifsyn committed:
to_year implementation

Laifsyn committed:
Expand tests

Laifsyn committed:
update container settings

Laifsyn committed:
Currency Implementation

Laifsyn committed:
Add Spanish implementation to lang::Language trait

Laifsyn committed:
try to ensure enum integrity safety

Laifsyn committed:
add rustfmt_skip

Laifsyn committed:
derive basic traits for Lang enum
The Lang enum should be really small since it doesn't contain any inherent data. So it should be dirt cheap to copy

Laifsyn committed:
Add more str parsing and invert bool logic

Laifsyn committed:
add Integration Test for spanish

Laifsyn committed:
Update Docs format and info

Laifsyn committed:
add Spanish Docs and fix currency logic

Laifsyn committed:
run Rustfmt on most files and add derives

Laifsyn committed:
testing DocTests

Laifsyn committed:
Try to improve Integration Test

Laifsyn committed:
Main became too much of a mess, so un-using it

Ballasi commented:
Hey, sorry for the big delays in replying to the messages!

I'll take a look at your PR soon, thanks for the changes.

I'll be honest by saying I am not much interested in dev containers, do you think you can get them removed for now?

Also I've seen that you've done a lot of improvements regarding fmt, that's nice! I am wondering if it's too hard for you to split that in another PR? Usually one PR = one feature!

Thanks again :)

Laifsyn commented:
oh yes. it's alright. 
As for the fmt it was something I ended up with after editing a template I copied and edited into my liking.

Laifsyn commented:
just... give me a moment while I figure how to separate the stuffs out

Laifsyn committed:
Remove devcontainers and rustfmt

Laifsyn committed:
attend clippy warning in test

Laifsyn commented:
I'll assume the rustfmt thing will got to be in a separate PR right?

Ballasi commented:
that would be best! just move files that are not changed in this PR to another one. If there was rustfmt ran within files that were changes within this PR, I am completely fine with keeping it here (on top of that, that'll avoid merge conflicts), makes it easier for both you and me :)

Laifsyn commented:
Sheesh. It makes me glad I didn't go crazy with changing stuffs. fortunately I found out about how to revert a file to a specific commit. Hope this is the way to go for such situation

Laifsyn committed:
undo rustfmt on bin.rs

Laifsyn commented:
Should I also undo the added Derives to the enums?

Laifsyn committed:
Undoing rustfmt on lang.rs

Laifsyn committed:
undo missed a rustfmt on lang.rs

Laifsyn committed:
undo rustfmt on lib.rs

Laifsyn committed:
try fix trailing whitespace

Laifsyn committed:
try fix EOF

Laifsyn committed:
delete unused main.rs

Laifsyn committed:
remove cfg attribute from num2words.rs

Laifsyn committed:
temporarily remove derives from enum Output in output.rs

Laifsyn committed:
Undo added derives to currency.rs

Laifsyn commented:
Alright, this should be about it. (except for mod.rs because I was too lazy to undo rustfmt on that. Shame on me)
But everything should be the most minimal for the spanish implementation feature.

Rustfmt should be easily be doable in another pull request by just re-adding the `rustfmt.toml` file and running rustfmt against the project.

the undonde derives could be additionally done in another PR.

and optionally fix clippy warnings in another PR

Laifsyn committed:
update README with new language

Laifsyn committed:
Update readme to reflect spanish as option

Laifsyn reviewed (commented):
I misread the commit content. This commit should've been `add missing docs`

Laifsyn committed:
temporarily remove derives

Ballasi commented:
Sorry for making you go through all of this trouble! and thank you very much for taking the time to work on it, it's very kind :)

I'll see the PR more in depth this weekend! If you're not planning on redoing the derives or rustfmt, I'll take on that job once this is merged.

Laifsyn commented:
I was thinking about figuring what rustfmt config was the one that onelined the chained methods and see if it was alright to disable it.

The reason is because if the method chaining gets one-lined, then the type-hint for each method return on the chain will not be hinted by rust-analyser.

but anyways, the chainings tends to be short more often than not, so the **loss of the type-hint** which can only be seen in an IDE isn't that relevant

Laifsyn commented:
Hi, I was thinking about making a [PR Benching with Criterion](https://github.com/Ballasi/num2words/pull/31) with another branch for benchmarks. Mainly  I want to see what issues can arises when I expect merge conflicts( expecting so because I had to add some derives to do the benchmark)

The reason was to see how the spanish implementation fared because I did a fix to the unneeded 0 spam that the split_thousands() method created. (Ref: [link](https://github.com/Ballasi/num2words/issues/11))

Ballasi commented on tests/lang_es_test.rs:
```diff
@@ -0,0 +1,88 @@
+    Year,
+    Currency,
+}
+fn to_words(num: BigFloat, output: Outputs, preference: &[&str]) -> Result<String, Num2Err> {
```
is there any reason why you are not using the actual builder each time, like on the other unit tests?

Laifsyn replied:
two actually.
1-) I wanted to keep the `assert` line into 1 line if I could
2-) It was a bit easier to see what changed between one line and the other. (ngl, the Closure was overkill just to skip 1 parameter but welp...)

Ballasi commented on tests/lang_es_test.rs:
```diff
@@ -0,0 +1,88 @@
+use num2words::{Currency, Lang, Num2Err, Num2Words};
```
any reasons why you split the tests in half and have this here? imho it should be all under es.rs

Laifsyn replied:
Edit: I misread

the `tests/*` folder is for integration tests

It simulates how the user would be using the library, and what output the user should expect.
Sort of similar to what the doc tests I wrote are for, but I felt too lazy to delete them after realising that they won't appear in the crate's docs

Tests inside `es.rs` would be to verify that internals (Which aren't public to outside the module) return the outputs that are expected. So if some shenanigans happens while linking stuffs to the public interface which the user will use, only the integration test can have possibility of breaking/failing. It legit is a waste of time, but I at the time (before figuring how integration tests sort of worked), I was more preoccupied with the internals instead of what the user would be getting

Disclaimer, I do however don't know if integration tests reflect in doc.rs. I believe they shouldn't

Ballasi commented on README.md:
```diff
@@ -83,6 +83,7 @@ Here is a list of all of the supported languages:
 | 🇫🇷🇨🇦 | `Lang::French`    | `fr`      | French      | quarante-deux |
 | 🇧🇪🇨🇩 | `Lang::French_BE` | `fr_BE`   | French (BE) | quarante-deux |
 | 🇨🇭   | `Lang::French_CH` | `fr_CH`   | French (CH) | quarante-deux |
+| 🇪🇸   | `Lang::Spanish`   | `es`      | Spanish     | cuarenta y dos|
```
do put Spanish above French! I am thinking of ranking that per total number of speakers, and Spanish has more people speaking it than French

also thinking adding the flag of Mexico after the Spanish flag would make sense

Laifsyn replied:
I originally put it below french because of alphabetic order. Curiously until now, popularity and alphabetic order matched lol
The mexican flag sounds alright. I just hope they don't have differing numbering (which remembers me, I need to see how to translate currency)

Ballasi commented on src/bin/bin.rs:
```diff
@@ -25,6 +25,7 @@ AVAILABLE LANGUAGES:
     fr:      French (France and Canada)
     fr_BE:   French (Belgium and the Democratic Republic of the Congo)
     fr_CH:   French (Swiss Confederation and Aosta Valley)
+    es:      Spanish
```
ditto

Laifsyn replied:
Ditto the pokemon?

Ballasi commented on src/lang/es.rs:
```diff
@@ -0,0 +1,1044 @@
+                }
+                // Boolean that checks if next Milliard is plural
+                let plural = triplet != 1;
+                match plural {
```
you should probably use a `if` case here rather than a `match` for bool (`if plural { ... } else { ... }`)

Laifsyn replied:
I figured it looked slimmer with match instead of if
```rust
// preview:
if plural {
    words.push(String::from(MILLARES[i]))
} else {
    words.push(String::from(MILLAR[i]))
}

// vs match
match plural {
    false => words.push(String::from(MILLAR[i])),
    true => words.push(String::from(MILLARES[i])),
}
```

Ballasi commented on tests/lang_es_test.rs:
```diff
@@ -0,0 +1,88 @@
+
+    let word = to_words(BigFloat::from(21.01), Outputs::Currency, &[]);
+    assert_eq!(word.unwrap(), "veintiún US dollars con un centavo");
+}
```
btw, compared the results from es.rs to python num2words and see differences

I don't mind much the differences in megas as it depends a lot on the source. however `mil trillones` makes me wonder, is this what should actually be the number?

```
10000000000000
diez billones
diez trillones

100000000000000
cien billones
cien trillones

1000000000000000000
un trillón
un quintillón

1000000000000000000000 <=== here
mil trillones
un sextillón

10000000000000000000000000
diez cuatrillones
diez septillones
```

Laifsyn replied:
What's different?. Is Mil Trillones from Python?

I followed the standard of Each thousands changes the Millard because It was a bit confusing that a Billion*1000 is called the equivalent in spanish of "thousand billions" instead of "trillion"

Laifsyn replied:
https://es.wikipedia.org/wiki/Anexo:Nombres_de_los_n%C3%BAmeros_en_espa%C3%B1ol
So according to this source, it starts to jump by 1_000_000 instead of 1_000 after the trillion

As summary, English's [Sextillion](https://en.wikipedia.org/wiki/English_numerals#Very_large_numbers) is 10^21
Spanish's Sextillion (according to [this particular page in wiki](https://es.wikipedia.org/wiki/Sextill%C3%B3n)) is 10^36

I guess it would still best to discuss with someone more familiarized with spanish, because if this really is the spanish standard, I'm pretty sure almost none will get spanish's quadrillon right (Because it actually is "A Million Trillion")

Ballasi commented on tests/lang_es_test.rs:
```diff
@@ -0,0 +1,88 @@
+
+    let word = to_words(BigFloat::from(21.01), Outputs::Currency, &[]);
+    assert_eq!(word.unwrap(), "veintiún US dollars con un centavo");
+}
```
also see differences in ordinals that are worth looking at:

```
12
décimosegundo (python num2words)
duodécimo (your implementation)

1000
milésimo
primer milésimo
```

Laifsyn replied:
"décimosegundo" could be just pyton's laziness to this edge case. But they're actually the same. From the Spanish Language academy: [duodécimo](https://dle.rae.es/duod%C3%A9cimo)

as for the 1000, I was really second guessing if ordinals also had the same rules as Cardinal when it regarded to the thousands. I'll probably add the edge case

Laifsyn replied:
also regarding the differences between `duodécimo` and `décimosegundo`, from quoted from https://www.rae.es/dpd/ordinales
`
c) Los ordinales correspondientes a los números 11 y 12 presentan hoy dos formas: las etimológicas simples undécimo y duodécimo, que son las preferidas en el uso culto, y las compuestas decimoprimero y decimosegundo, creadas modernamente por analogía con la forma que adoptan los ordinales del resto de las series (vigesimoprimero, vigesimosegundo, trigésimo primero, trigésimo segundo, etc.).
`
So as  a summary, the biword terms of `decimoprimero` and `decimosegundo` are just terms that has been come up with as an analogy to the rest of the ordinals expression

Ballasi reviewed (commented):
Thank you again for taking the time to participate! I'm sorry I am relatively slow, I've got a lot going on lately.

Feel free to take a look at the comments I've put :)

Laifsyn committed:
Fix Tens edge case for Ordinals

based on [Real Academia Española](https://www.rae.es/dpd/ordinales) standards, the TENS of 2 have no space between itself and its units.
`"la primera y a la segunda decena se pueden escribir en una o en dos palabras, siendo hoy mayoritaria y siempre preferible la grafía univerbal"`

Laifsyn committed:
Fix Ordinals Representation
From https://github.com/Ballasi/num2words/pull/29#discussion_r1565124539,
fix ordinal representation based on the rules defined in 2.b && 2.d @
https://www.rae.es/dpd/ordinales

Laifsyn commented:
Alright. That should be it

Laifsyn committed:
Fix Tests & remove comments

Laifsyn commented:
Oh, btw. I'm working on translating the currencies. Might submit a commit by the end of the day

Laifsyn committed:
Currency Translation for spanish

Laifsyn committed:
Cardinal Fix - Mistakenly used English semantics (#1)

Fix the wrong semantics that was used.
Spanish semantics follows this pattern of for each big milliard (Million, Billion, Trillion) grows at a pace of `1 000 000`^n

This means that 1 million is `1 000 000`^1
1 billion => `1 000 000`^2 [1 000_000 000_000]
1 trillion => `1 000 000`^3 [1 000_000 000_000 000_000]
1 quadrillion => `1 000 000`^4 [......]
...... etc

This semantic differs from english's
1 billion => `1 000`^3 [1 000 000_000] 
1 trillion => `1 000 000`^4 [1 000_000 000_000]
1 quadrillion => `1 000 000`^4 [......]

Laifsyn committed:
Fix tests due to incorrect Milliard Semantic

Laifsyn committed:
remove unreachable code

Laifsyn commented:
I have a question if you can get the answer

I'll be leaving it as the first meanwhile, but should it be like the second case?

```rust
assert_eq!(
    ordinal(124_001_091_000_000_000_001),
    "ciento veinticuatrotrillonésimos milnoventa y unobillonésimos primeros"
//  "ciento veinticuatrotrillonésimos mil noventa y unobillonésimos primeros"
);
```

Laifsyn committed:
Ordinal_fix (#3)

* Staging changes for ordinal Fix

Some data to take into consideration
Todo: Fix accent on monomorphized words i.e. VigesimoSegundo

Si el ordinal se escribe en dos palabras, el primer elemento mantiene la tilde que le corresponde como palabra independiente: vigésimo segundo, vigésima cuarta, trigésimo octavo, cuadragésima quinta; pero, si se escribe en una sola palabra, el compuesto resultante, al ser ser una voz llana terminada en vocal, debe escribirse sin tilde, pues no le corresponde llevarla según las reglas de acentuación (v. cap. II, § 3.4.5.1.1): vigesimosegundo (no ⊗‍vigésimosegundo).

Los ordinales complejos escritos en una sola palabra solo presentan variación de género y número en el segundo componente: vigesimoprimero, vigesimoprimera, vigesimoprimeros, vigesimoprimeras; pero, si se escriben en dos palabras, ambos componentes son variables: vigésimo primero, vigésima primera, vigésimos primeros, vigésimas primeras. No se consideran correctas las grafías en dos palabras si se mantiene invariable el primer componente: ⊗‍vigésimo segundos, ⊗‍vigésimo cuarta, ⊗‍vigésimo octavas.

* temporarily add rustfmt back

* Finish Ordinal Fix.

Reference: https://www.rae.es/dpd/ordinales

* remove accent on composites of `Vigesimo`

* remove rustfmt for merge

Laifsyn commented:
I couldn't redo the Cardinal implementation.
Performance gains were dirt low. at about x1.02-x1.25 faster, and also the implementation looked a bit ugly. So I have no changes to the current implementation of Spanish.

Laifsyn committed:
Update split_thousand to match other langs implementation

PabloGmz96 commented:
Hey guys! Native Spanish speaker here. Is there anything that you need help with? I'd love to use this crate in a project that I'm currently developing

Laifsyn commented:
> Hey guys! Native Spanish speaker here. Is there anything that you need help with? I'd love to use this crate in a project that I'm currently developing
- [ ] I would like opinions on whether to support a flavor for "Semantically Incorrect" milliards. 
according to  RAE, 10E9(1,000,000,000) is called "Billón", whereas in English, 10E9 is called "Trillion". Choosing to support semantically Incorrect cardinals would mean that ~~spanish's "Billón" would be english's trillion~~ milliards like "billón" and "trillón" in spanish are just the literal translation of english's.

- [ ] Maybe check currency, in case some currency got mistranslated.


- [ ] ~~Support converting too big of numbers into `n`E`m` expression, i.e. `20.23E20` => `veinte coma veintitrés por 10 a la 20`~~

Priority close to 0:
* Cleanup tests & Integration Tests: Because some test cases overlaps other tests. Also I personally think they could look cleaner with some work.

Otherwise, everything should be fine. Other than big numbers which can lose precision

PabloGmz96 commented:
Hey @Laifsyn! I just review the "Semantically Incorrect" numbers that you mention and as far as I know almost the complete Latin-america region use those big numbers exactly in the same way as in English, so a direct translation of them should be precise enough. Later versions of this crate could include some other forms of Spanish cardinals numbers but knowing that they are rarely used this should be not a priority. But let me ask some friends in Colombia and Argentina (Countries that use particularly high number prices in the every day life) to confirm that.

In the other hand, let me test the USD and PESOS translation, those are the currencies that I'm familiar with.

And last but not least, I just want to say a big thank you to make the Spanish implementation possible!

PabloGmz96 commented:
Just a quick question... I'm using this crate in library mode, is there any way to just print the resulting words? The output always show me the array that the code use to convert numbers to words.
![image](https://github.com/Ballasi/num2words/assets/76931310/4ad6c270-529e-4060-b19c-b80a26f2c818)

PabloGmz96 commented:
Just a quick update... When using Spanish as language, the output shows the array that the code use to translate the numbers to words.
![image](https://github.com/Ballasi/num2words/assets/76931310/3e1a7955-992f-44f7-b4ee-e0e2673b7df9)

Laifsyn committed:
Remove unused println! call

Laifsyn commented:
> Just a quick update... When using Spanish as language, the output shows the array that the code use to translate the numbers to words. ![image](https://private-user-images.githubusercontent.com/76931310/341352048-3e1a7955-992f-44f7-b4ee-e0e2673b7df9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4OTA1OTMsIm5iZiI6MTcxODg5MDI5MywicGF0aCI6Ii83NjkzMTMxMC8zNDEzNTIwNDgtM2UxYTc5NTUtOTkyZi00NGY3LWI0ZWUtZTBlMjY3M2I3ZGY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIwVDEzMzEzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyODViN2FmMjhmOGMxMjE3YmIzYmYzNDFiYWM0MTU4OGViYjliODRkMDQzNDYyMjQ5NDQ2Mjk3NGEyYzNhNTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._0d4u5XH2EiQmF2Lp2IZx_eyOc10Rr49QaW1IB8XYz8)

Thanks for finding out! I might have forgotten to test the actual binary before I uploaded

bd-g commented:
Any chance on getting this merged and pushed out?

PabloGmz96 commented:
I just want to add that a year after PROD usage of this branch, I've found zero issues in Cardinal translation!

Laifsyn commented:
> I just want to add that a year after PROD usage of this branch, I've found zero issues in Cardinal translation!

Wait, so it means there were issues in other places? 💀
